crates/red_knot_python_semantic/src/semantic_index/use_def/bitset.rs

```diff
@@ -0,0 +1,402 @@
+                    myblocks[i] &= other_blocks[i];
+                }
+            }
+            (Self::Heap(myset), Self::Heap(other_set)) => {
+                *myset = myset.intersection(other_set).copied().collect();
+            }
+            (me, other) => {
+                for value in other.iter() {
+                    me.insert(value);
+                }
+            }
+        }
+    }
+
+    /// Return an iterator over the values (in ascending order) in this [`BitSet`].
+    pub(super) fn iter(&self) -> BitSetIterator<'_, B> {
+        match self {
+            Self::Inline(blocks) => BitSetIterator::Inline {
+                blocks,
+                cur_block_index: 0,
+                cur_block: blocks[0],
+            },
+            Self::Heap(set) => BitSetIterator::Heap(set.iter()),
+        }
+    }
+}
+
+/// Iterator over values in a [`BitSet`].
+#[derive(Debug)]
+pub(super) enum BitSetIterator<'a, const B: usize> {
+    Inline {
+        /// The blocks we are iterating over.
+        blocks: &'a [u64; B],
+
+        /// The index of the block we are currently iterating through.
+        cur_block_index: usize,
+
+        /// The block we are currently iterating through (and zeroing as we go.)
+        cur_block: u64,
+    },
+    Heap(btree_set::Iter<'a, u32>),
+}
+
+impl<const B: usize> Iterator for BitSetIterator<'_, B> {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Inline {
+                blocks,
+                cur_block_index,
+                cur_block,
+            } => {
+                while *cur_block == 0 {
+                    if *cur_block_index == B - 1 {
+                        return None;
+                    }
+                    *cur_block_index += 1;
+                    *cur_block = blocks[*cur_block_index];
+                }
+                let value = cur_block.trailing_zeros();
+                // reset the lowest set bit
+                *cur_block &= cur_block.wrapping_sub(1);
+                // SAFETY: `value` cannot be more than 64, `cur_block_index` cannot be more than
+                // `B - 1`, and we check above that `B * 64 < u32::MAX`. So both `64 *
+                // cur_block_index` and the final value here must fit in u32.
+                #[allow(clippy::cast_possible_truncation)]
+                Some(value + (64 * *cur_block_index) as u32)
+            }
+            Self::Heap(set_iter) => set_iter.next().copied(),
+        }
+    }
+}
+
+impl<const B: usize> std::iter::FusedIterator for BitSetIterator<'_, B> {}
+
+/// Array of [`BitSet<B>`]. Up to N stored inline, more than that overflows to a Vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum BitSetArray<const B: usize, const N: usize> {
+    Inline {
+        /// Array of N [`BitSet`].
+        array: [BitSet<B>; N],
+
+        /// How many of the bitsets are used?
+        size: usize,
+    },
+
+    Heap(Vec<BitSet<B>>),
+}
```

Comment:
Yeah I knew about smallvec, not sure why I didn't use it for `BitSetArray`. I'll try that and see if it works; I've had lifetime-invariance problems before when I've tried using smallvec.

Reply:
Yeah, you can run into lifetime errors with smallvec because the variance is incorrect https://github.com/servo/rust-smallvec/issues/146 I'm desperately waiting for small vec 2

Reply:
This is a great suggestion too; no lifetime errors, no effect on perf, but allows removing a lot of code.